Switch APIs for semantic classification in roslyn to take TextSpan[]

maryamariyan opened the pull request:
Alternative to https://github.com/maryamariyan/roslyn/pull/1

As with latest in commit https://github.com/maryamariyan/roslyn/pull/2/commits/00ca9109cbb6afc5e081d619a20bcc1d1e5c43f5:
![image](https://github.com/maryamariyan/roslyn/assets/5897654/b27ec392-47e5-4c99-97c2-8ce9e8628ef2)

This experiment is done by reopening the index.razor file in aspire repo multiple times (but first made the file 10X larger and zoomed out to 40% so it would include more items in the span) and then doing a side by side comparison:

- the `s_old` indicates when we send only a single span without this PR.
- the `s_new` indicates when we send down multiple spans in the current PR.

Rekkonnect committed:
Add comment

Rekkonnect committed:
Adjust checks for pattern expression context

Rekkonnect committed:
Improve pattern evaluation logic + tests

Rekkonnect committed:
Move constant pattern context processing

CyrusNajmabadi committed:
Merge remote-tracking branch 'upstream/main' into fix-70226

CyrusNajmabadi committed:
Simplify

CyrusNajmabadi committed:
Simplify

CyrusNajmabadi committed:
In progress

CyrusNajmabadi committed:
In progress

CyrusNajmabadi committed:
in progress

CyrusNajmabadi committed:
First pass done

CyrusNajmabadi committed:
Working test

CyrusNajmabadi committed:
porting tests

CyrusNajmabadi committed:
All tests

CyrusNajmabadi committed:
In progress

CyrusNajmabadi committed:
Merge remote-tracking branch 'upstream/main' into convertToRegularConstructor

CyrusNajmabadi committed:
Narrow the search

CyrusNajmabadi committed:
Fixes

CyrusNajmabadi committed:
Update nested type references

CyrusNajmabadi committed:
in progress

CyrusNajmabadi committed:
Fixes

CyrusNajmabadi committed:
inner node

CyrusNajmabadi committed:
in progrss

CyrusNajmabadi committed:
Update xml

CyrusNajmabadi committed:
in proress

CyrusNajmabadi committed:
move all initializers

CyrusNajmabadi committed:
Handle multiple initializers case

CyrusNajmabadi committed:
Cleaner doc code

CyrusNajmabadi committed:
Merge remote-tracking branch 'upstream/main' into convertToRegularConstructor

CyrusNajmabadi committed:
Doc comments

CyrusNajmabadi committed:
Fies

CyrusNajmabadi committed:
further along

CyrusNajmabadi committed:
Fixes

CyrusNajmabadi committed:
Workign docs

CyrusNajmabadi committed:
More fixups

CyrusNajmabadi committed:
Do not know what's going on here

CyrusNajmabadi committed:
Cleanup

CyrusNajmabadi committed:
Move to new file

CyrusNajmabadi committed:
Inlined

CyrusNajmabadi committed:
In progress

CyrusNajmabadi committed:
inline

CyrusNajmabadi committed:
cleanup

CyrusNajmabadi committed:
simplify

CyrusNajmabadi committed:
Make readonly

CyrusNajmabadi committed:
Fix indentation

CyrusNajmabadi committed:
Rename tests

CyrusNajmabadi committed:
simplify

CyrusNajmabadi committed:
indent

CyrusNajmabadi committed:
Mark diagnostics as being 'live' or not in LSP

CyrusNajmabadi committed:
Fix docs

CyrusNajmabadi committed:
Clear 'run code analysis' results when a build completes

CyrusNajmabadi committed:
lint

CyrusNajmabadi committed:
Clear things out

CyrusNajmabadi committed:
Cleanup

CyrusNajmabadi committed:
Docs

CyrusNajmabadi committed:
Fix classification layering

CyrusNajmabadi committed:
Merge remote-tracking branch 'upstream/main' into fix-70226

mavasani committed:
Handle missing ISimplifactionService implementation for non-C#/VB langauges

Fixes [AB#1901118](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1901118)

Verified that the added unit test fails with the same exception as in the above bug prior to the product change.

Cosifne committed:
Update config file after 17.9 branch is created

CyrusNajmabadi committed:
Update breaking change docs

genlu committed:
Fix option used in test

CyrusNajmabadi committed:
tweak

CyrusNajmabadi committed:
Update docs/Breaking API Changes.md

CyrusNajmabadi committed:
Update docs/Breaking API Changes.md

CyrusNajmabadi committed:
Merge pull request #70435 from CyrusNajmabadi/updateBCs

Update breaking change docs

CyrusNajmabadi committed:
Minor cleanup

jasonmalinowski committed:
Actually set the project display name

jasonmalinowski committed:
Consolidate all the calls to Distinct()

CyrusNajmabadi committed:
Be resilient to failure case

CyrusNajmabadi committed:
SImplify results

dotnet-bot committed:
Merge pull request #70441 from dotnet/merges/release/dev17.8-to-main

Merge release/dev17.8 to main

CyrusNajmabadi committed:
Update test harness

CyrusNajmabadi committed:
Update test harness

CyrusNajmabadi committed:
Flow data along

CyrusNajmabadi committed:
further along

CyrusNajmabadi committed:
REstore logic

CyrusNajmabadi committed:
Closer

genlu committed:
TriggerOnTyping should not block explicit invoke

CyrusNajmabadi committed:
Fixups

genlu committed:
Move some trigger logic from AsyncCompletion to CompletionService

CyrusNajmabadi committed:
Fix test

CyrusNajmabadi committed:
move to bottom

CyrusNajmabadi committed:
go back

genlu committed:
Undo change in vb test

CyrusNajmabadi committed:
reorder

jasonmalinowski committed:
Log some info about the project to the trace log

Just a general improvement: when we're loading projects, output a
count of the various things loaded, so obvious problems can be found.

CyrusNajmabadi committed:
Apply suggestions from code review

genlu committed:
Merge remote-tracking branch 'dotnet/main' into FixCompletionTrigger

jasonmalinowski committed:
Localize a bunch of user-visible strings

jasonmalinowski committed:
Log when a project is received from Dev Kit

dibarbet committed:
Ensure the VSCode server process runs on non-preview runtime

ToddGrun committed:
Slight improvement to refactoring preview performance for InlineTemporaryCodeRefactoringProvider (#70440)

The GetSemanticModelAsync calls are expensive now as they run source generators when the document they are being requested on doesn't yet have a compilation.

jjonescz committed:
Rename Razor compiler assemblies (#69655)

* Rename Razor compiler assemblies

* Fixup a test

* Consider all past and future razor DLL names

dibarbet committed:
Update src/Workspaces/Core/MSBuild.BuildHost/Program.cs

Co-authored-by: Jason Malinowski <jason@jason-m.com>

dibarbet committed:
Update src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs

Co-authored-by: Jason Malinowski <jason@jason-m.com>

CyrusNajmabadi committed:
Merge pull request #70443 from CyrusNajmabadi/emPerf

Refactor how Extract Method returns results.

CyrusNajmabadi committed:
Simplify

CyrusNajmabadi committed:
Simplify

CyrusNajmabadi committed:
remove loc string

CyrusNajmabadi committed:
speeling

CyrusNajmabadi committed:
move switch

CyrusNajmabadi committed:
Simplify

dibarbet committed:
Merge pull request #70446 from dibarbet/remove_runtime_framework_version

Ensure the VSCode server process runs on non-preview runtime

genlu committed:
Add a test

genlu committed:
Filter completion list only with text before cursor

CyrusNajmabadi committed:
Move checks earlier

CyrusNajmabadi committed:
Fixes

jasonmalinowski committed:
Merge pull request #70157 from jasonmalinowski/bump-structured-logger

Bump MSBuild.StructuredLogger package version

mavasani committed:
Merge pull request #70425 from mavasani/HandleMissingSimplificationService

Handle missing ISimplificationService implementation for non-C#/VB languages

CyrusNajmabadi committed:
Merge pull request #70399 from CyrusNajmabadi/convertToRegularConstructor

Add refactoring to convert a primary-constructor-parameter to a normal constructor.

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/source-build-externals build 20231018.1 (#70454)

[main] Update dependencies from dotnet/source-build-externals

jjonescz committed:
Revert "[main] Update dependencies from dotnet/arcade (#70270)" (#70451)

This reverts commit bcdf3fd2d893ed8c72165fe534e91efc671099e1.

CyrusNajmabadi committed:
Merge pull request #70412 from CyrusNajmabadi/clearRCA

Clear 'run code analysis' results when a build completes

CyrusNajmabadi committed:
Merge pull request #70262 from Rekkonnect/fix-70226

Improve completion in the context of pattern expressions

ToddGrun committed:
Prevent stringbuilder resizes during SyntaxNode.GetText (#70447)

In the profile I'm looking at, CSWin32.SourceGenerator is the primary caller of this method, but the extractmethod coderefactoring provider also works it's way into this code.

mgravell committed:
docs - note that interceptors are C# only (#70457)

CyrusNajmabadi committed:
in progress

genlu committed:
Merge pull request #70448 from genlu/FixFiltering

Filter completion list only with text before cursor

CyrusNajmabadi committed:
Not quite

CyrusNajmabadi committed:
almost

genlu committed:
Merge pull request #70444 from genlu/FixCompletionTrigger

Fix TriggerOnTyping to not block explicit invoking completion

CyrusNajmabadi committed:
Unify selection logic

CyrusNajmabadi committed:
Unify selection logic

CyrusNajmabadi committed:
front load hecks

CyrusNajmabadi committed:
synchronous

CyrusNajmabadi committed:
Move method

CyrusNajmabadi committed:
Fi

CyrusNajmabadi committed:
Fix tests

CyrusNajmabadi committed:
Fix tests

CyrusNajmabadi committed:
Fix tests

CyrusNajmabadi committed:
Merge remote-tracking branch 'upstream/main' into emPerf2

CyrusNajmabadi committed:
make nullable

Cosifne committed:
Merge pull request #70442 from Cosifne/dev/shech/create17.9Branch

Update config file after 17.9 branch is created

CyrusNajmabadi committed:
Nullable

dibarbet committed:
Ensure diagnostics are logged before proceeding with further project loading

CyrusNajmabadi committed:
Simplify

CyrusNajmabadi committed:
Delete

CyrusNajmabadi committed:
remove

CyrusNajmabadi committed:
remove

tmat committed:
Fix incorrect contract in VB EnC analyzer (#70464)

CyrusNajmabadi committed:
cleanup

CyrusNajmabadi committed:
Delete unnecessary code

CyrusNajmabadi committed:
Delete unnecessary code

CyrusNajmabadi committed:
Almost there

CyrusNajmabadi committed:
Fixes

CyrusNajmabadi committed:
Fix script

CyrusNajmabadi committed:
simplify

CyrusNajmabadi committed:
simplify

cston committed:
Collection expressions: spread elements are not expressions (#70417)

CyrusNajmabadi committed:
In progress

CyrusNajmabadi committed:
Simplify

CyrusNajmabadi committed:
Apply suggestions from code review

CyrusNajmabadi committed:
move

CyrusNajmabadi committed:
Merge branch 'emPerf2' of https://github.com/CyrusNajmabadi/roslyn into emPerf2

CyrusNajmabadi committed:
lint

dibarbet committed:
Ensure evaluation results are included in binlog

CyrusNajmabadi committed:
Fix issue where primary constructors were not being grouped across TFMs in the results window

CyrusNajmabadi committed:
move comment

CyrusNajmabadi committed:
Move

CyrusNajmabadi committed:
lint

CyrusNajmabadi committed:
Simplify construction

CyrusNajmabadi committed:
defer computation

CyrusNajmabadi committed:
defer computation

CyrusNajmabadi committed:
in progrss

CyrusNajmabadi committed:
in progrss

CyrusNajmabadi committed:
Merge pull request #70476 from CyrusNajmabadi/farPrimaryConstructosr

Fix issue where primary constructors were not being grouped across TFMs in the results window

CyrusNajmabadi committed:
vb and c# sides

CyrusNajmabadi committed:
Working

CyrusNajmabadi committed:
simplify

CyrusNajmabadi committed:
simplify

CyrusNajmabadi committed:
Small cleanup

DoctorKrolic committed:
Optimize `T? ?? <expr>` to `T.GetValueOrDefault(<expr>)` for a side-effectless `<expr>` where `T?` is a nullable value type (#70199)

Closes #56007

Commit:
Update ExternalAccess for EditorConfig (#70415)

* move editorconfiggenerator to external access

* fix reference

* add interface

* fix test reference

* add key

* fix tests

* add to vsix?

* Pr comments

* revert tests

* final pr cleanup

CyrusNajmabadi committed:
Merge pull request #70466 from CyrusNajmabadi/emPerf2

Front load extract-method syntax and semantic checks.

CyrusNajmabadi committed:
Merge branch 'main' into emPerf3

CyrusNajmabadi committed:
Merge pull request #70477 from CyrusNajmabadi/emPerf3

Defer extract method mutation computation till the point the code action is invoked.

CyrusNajmabadi committed:
Merge pull request #70480 from CyrusNajmabadi/syntaxCleanup

Tiny cleanup in SyntaxGenerator.

CyrusNajmabadi committed:
Merge pull request #70402 from CyrusNajmabadi/liveSource

Mark diagnostics as being 'live' or not in LSP

CyrusNajmabadi committed:
Fix 'use collection expression' with await-foreach calls

dibarbet committed:
Skip elements with false conditions during evaluation
This fixes issues in some projects where the evaluation of the element
will fail if the condition does not hold

Commit:
update pipe name (#70492)

dibarbet committed:
Include more settings to better match VS behavior

jjonescz committed:
Prepare for merging razor tests (#70485)

jcouv committed:
Validate diagnostic ID in Experimental attribute (#70397)

ToddGrun committed:
Allow invisible editor support to handle open aspx files. (#70396)

* Allow invisible editor support to handle open aspx files.

The issue here is that the aspx docdata doesn't implement IVsTextLines directly, but implements IVsTextBufferProvider as a mechanism to obtain the text lines. Added code to fallback to QI as such.

Addresses https://github.com/dotnet/roslyn/issues/67410

CyrusNajmabadi committed:
Remove unnecessary casts

CyrusNajmabadi committed:
Remove unnecessary casts

dibarbet committed:
Remove record properties which we do not use

dibarbet committed:
Add tests to verify logger is called

dibarbet committed:
Add comments

dibarbet committed:
Update integration test to expect new behavior

dibarbet committed:
Ensure execution diagnostics are logged as well

jjonescz committed:
Skip VerifyOverloadsUpdated integration tests

jjonescz committed:
Publish EA.EditorConfigGenerator (#70510)

dibarbet committed:
Fix another integration test to reflect new behavior

jjonescz committed:
Handle generic types as parameters of attributes in `MetadataDecoder` (#70151)

* Handle generic types as parameters of attributes in `MetadataDecoder`

* Add more tests

* Revert handling generic type parameters

This reverts commit d280b1e94b879b24f8beacdccab984d94bb09da8
except tests.

* Use parameter types from constructor symbol

* Avoid null lists when attribute constructor is null

* Add more VB tests

* Update a test

* Handle `System.Type`

* Check that attribute constructor returns void

* Use `IParameterSymbolInternal`

* Add `IArrayTypeSymbolInternal`

* Fixup NoPIA tests

* Remove `catch`

* Fix extension invocation

* Change target framework of just the PIA test

* Check that attribute constructor is not generic

* Simplify VB tests

* Revert changes to NoPIA tests

Reverts 27a167ead2ca8db393e89f2181252ffd0244dc6d and f44340e733acad00dae3c7e795203591b52d17ae.

* Crack PIA attributes from metadata

* Use explicit types

* Use existing method for Guid attribute

* Capture PIA attributes during decoding

* Crack PIA attributes from metadata in VB

* Remove unused `out` vars

* Use GUID string from `DecodeGuidAttribute`

alrz committed:
Utilize ConstantValueOne while emitting conditionals (#70352)

Commit:
Merge pull request #70512 from jjonescz/70511-SkipVerifyOverloadsUpdated

Skip VerifyOverloadsUpdated integration tests

dibarbet committed:
Merge remote-tracking branch 'upstream/main' into fix_evaluation_skip

dibarbet committed:
Merge remote-tracking branch 'upstream/main' into fix_error_logging

dibarbet committed:
Merge remote-tracking branch 'upstream/main' into binlog_evaluation

CyrusNajmabadi committed:
Merge remote-tracking branch 'upstream/main' into awaitCollectionExpr

CyrusNajmabadi committed:
Merge remote-tracking branch 'upstream/main' into unnecessaryCasts

dotnet-bot committed:
Localized file check-in by OneLocBuild Task: Build definition ID 327: Build ID 2299291

dotnet-bot committed:
Localized file check-in by OneLocBuild Task: Build definition ID 327: Build ID 2299291

jasonmalinowski committed:
#nullable enable RelativePathResolver

jasonmalinowski committed:
Resolve reference paths to absolute paths

Fixes https://github.com/dotnet/vscode-csharp/issues/6584

CyrusNajmabadi committed:
Merge pull request #70504 from CyrusNajmabadi/unnecessaryCasts

Remove unnecessary casts.

jasonmalinowski committed:
Attempt to work around GitHub Copilot breaking integration tests

When the GitHub Copilot extension launches, it may create an output
window and asynchronously steal focus. This attempts to force it to load
at the start of VS and then waits a bit in hopes we're past the point
where it'll steal focus.

CyrusNajmabadi committed:
Fix flakeyness in extract method test

CyrusNajmabadi committed:
NRT

Commit:
Merge pull request #70541 from dotnet/dev/arkalyan/classificationlayering

Cherry pick: Fix classification layering

Commit:
Merge pull request #70526 from dotnet/locfiles/3431acd8-41e3-4b96-8c84-02db235d1513

Localized file check-in by OneLocBuild Task: Build definition ID 327: Build ID 2300362

jasonmalinowski committed:
Merge pull request #70472 from dibarbet/binlog_evaluation

Ensure evaluation results are included in binlog

jasonmalinowski committed:
Merge pull request #70532 from jasonmalinowski/attempt-to-work-around-github-copilot-issue

Attempt to work around GitHub Copilot breaking integration tests

jasonmalinowski committed:
Merge pull request #70467 from dibarbet/fix_error_logging

Ensure diagnostics are logged before proceeding with further project loading

CyrusNajmabadi committed:
Merge remote-tracking branch 'upstream/main' into emFlakey

CyrusNajmabadi committed:
Merge remote-tracking branch 'upstream/main' into awaitCollectionExpr

dibarbet committed:
Merge pull request #70496 from dibarbet/fix_evaluation_skip

Skip elements with false conditions during evaluation

jjonescz committed:
Merge branch 'release/dev17.8' into main

jjonescz committed:
Fix character offset off-by-one error in enhanced line directives (#70247)

* Add a test

* Fix character offset off-by-one error in enhanced line directives

* Update tests

ToddGrun committed:
Reduce dependency on ImmutableDictionary in CompletionItem's properties storage. (#70474)

ImmutableDictionary is not the right data structure for CompletionItem properties for several reasons.

1) I've seen the property immutable dictionary creation/modification show up in profiles
2) From the testing I've done, the properties collection hasn't yet exceeded 5 items

Because these property collections are so small, just using an ImmutableArray for them would be significantly faster for creation, modification, and querying. The biggest problem I see with changing this is that the property collection is exposed as a public property on CompletionItem. As that can't be changed, I instead had that property be delay populated, with no Roslyn codepaths accessing that dictionary.

I tested with and without this change by bringing up completion 10 times within a single file with no edits between. For this scenario, I saw the following changes:

CPU: 1658ms -> 1424 ms
Allocations: 158 MB -> 137 MB

CyrusNajmabadi committed:
Ensure we cache both the last semantic and last embedded-lang classification for a document

CyrusNajmabadi committed:
Remove parallel

CyrusNajmabadi committed:
Merge pull request #70494 from CyrusNajmabadi/awaitCollectionExpr

Fix 'use collection expression' with await-foreach calls

Commit:
Merge pull request #70543 from CyrusNajmabadi/emFlakey

Fix flakeyness in extract method test

AlekseyTs committed:
Clean up inheritance hierarchy for CSharpAttributeData/VisualBasicAttributeData. (#70519)

This is a refactoring in preparation for more changes related to #70338.
- Break inheritance between SynthesizedAttributeData/RetargetingAttributeData and SourceAttributeData.
- Strengthen detection of well-known attributes for RetargetingAttributeData
- Remove redundant inputs for some APIs

CyrusNajmabadi committed:
Handle authorization exceptions when reading files

CyrusNajmabadi committed:
Fix

CyrusNajmabadi committed:
wider scope

CyrusNajmabadi committed:
Cleanup

CyrusNajmabadi committed:
Merge pull request #70550 from CyrusNajmabadi/classificationCacheFix

Ensure we cache both the last semantic and last embedded-lang classification for a document

jasonmalinowski committed:
Apply the GitHub Copilot workaround to new-framework integration tests

jasonmalinowski committed:
Merge pull request #70561 from jasonmalinowski/apply-github-copilot-workaround

Apply the GitHub Copilot workaround to new-framework integration tests

jasonmalinowski committed:
Merge pull request #70527 from jasonmalinowski/fully-qualify-reference-paths

maryamariyan committed:
Switch API in roslyn to take TextSpan[]

maryamariyan commented on src/Workspaces/Core/Portable/Classification/SyntaxClassification/AbstractSyntaxClassificationService.Worker.cs:
```diff
@@ -99,9 +100,15 @@ private void ProcessNodes()
+                        var intervalsThatIntersect = _textSpanIntervalTree.GetIntervalsThatIntersectWith(
+                            nodeOrToken.FullSpan.Start, nodeOrToken.FullSpan.Length);
+
+                        foreach (var interval in intervalsThatIntersect)
```
do we want to call `ClassifyNodeOrToken` for all intervals or only first one is enough

ToddGrun replied:
I would think all of them. One question I have here is whether you need to call both HasIntervalThatIntersectsWith and GetIntervalsThatIntersectWith. I would think you could just call GetIntervalsThatIntersectWith and if it's Length > 0, then go in and do the work

ToddGrun commented on src/Features/Core/Portable/EmbeddedLanguages/Classification/AbstractEmbeddedLanguageClassificationService.cs:
```diff
@@ -4,13 +4,14 @@
 
 using System;
 using System.Collections.Generic;
+using System.Collections.Immutable;
```
is this used somewhere?

ToddGrun commented on src/Workspaces/Core/Portable/Classification/AbstractClassificationService.cs:
```diff
@@ -159,13 +159,16 @@ private static bool IsFullyLoaded(Document document, CancellationToken cancellat
-                    var reassignedVariableSpans = await reassignedVariableService.GetLocationsAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
-                    foreach (var span in reassignedVariableSpans)
-                        result.Add(new ClassifiedSpan(span, ClassificationTypeNames.ReassignedVariable));
+                    for (var i = 0; i < textSpans.Length; i++)
```
>for (var i = 0; i < textSpans.Length; i++) [](http://example.com/codeflow?start=20&length=42)

just foreach

ToddGrun commented on src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs:
```diff
@@ -45,12 +46,15 @@ internal static partial class ClassifierHelper
             using var _2 = Classifier.GetPooledList(out var semanticSpans);
 
-            await classificationService.AddSyntacticClassificationsAsync(document, span, syntaxSpans, cancellationToken).ConfigureAwait(false);
+            for (var i = 0; i < spans.Length; i++)
```
>for (var i = 0; i < spans.Length; i++) [](http://example.com/codeflow?start=12&length=38)

Did this one not lend itself to passing down the array like the others?

maryamariyan replied:
`Worker.vb` and `Worker.cs` would need to change in order to make `AddSyntacticClassifications` work (taking array), I will first get a profile before changing this so we would know if it is needed at all

CyrusNajmabadi replied:
i would pass this along.  The *inner* call can do thsi loop.

ToddGrun commented on src/Workspaces/Core/Portable/Classification/SyntaxClassification/AbstractSyntaxClassificationService.Worker.cs:
```diff
@@ -44,7 +45,7 @@ internal partial class AbstractSyntaxClassificationService
                 _semanticModel = semanticModel;
                 _syntaxTree = semanticModel.SyntaxTree;
-                _textSpan = textSpan;
+                _textSpanIntervalTree = new TextSpanIntervalTree(textSpans);
```
>new TextSpanIntervalTree(textSpans); [](http://example.com/codeflow?start=40&length=36)

As mentioned earlier, I'm not sure of the costs of creating this. You probably should verify in a profile if this is expensiive, and if so, share the one used here and in the semantic classifier

maryamariyan committed:
Apply PR feedback

mavasani committed:
Workaround issue #70570

Short term workaround for issue #70570 by moving IDE00555 warning configuration entry from globalconfig to editorconfig. Verified that IDE0055 warnings are now getting reported locally on the repo.

ToddGrun commented on src/Workspaces/Core/Portable/Classification/SyntaxClassification/AbstractSyntaxClassificationService.Worker.cs:
```diff
@@ -98,10 +99,15 @@ private void ProcessNodes()
-                        ClassifyNodeOrToken(nodeOrToken);
+                        foreach (var interval in intervalsThatIntersect)
+                        {
+                            ClassifyNodeOrToken(nodeOrToken, interval);
```
>ClassifyNodeOrToken [](http://example.com/codeflow?start=28&length=19)

It looks like there is something a little tricky here. If this works it's way down into ClassifyStructuredTrivia, each iteration of this loop could add the same trivia structure into _pendingNodes.

mavasani committed:
Merge pull request #70571 from mavasani/Workaround_70570

Workaround issue #70570

dotnet-maestro[bot] committed:
[main] Update dependencies from dotnet/arcade (#70482)

* Update dependencies from https://github.com/dotnet/arcade build 20231018.2

Microsoft.DotNet.Arcade.Sdk , Microsoft.DotNet.Helix.Sdk
 From Version 9.0.0-beta.23504.4 -> To Version 9.0.0-beta.23518.2

Dependency coherency updates

Microsoft.DotNet.XliffTasks
 From Version 1.0.0-beta.23502.1 -> To Version 1.0.0-beta.23516.1 (parent: Microsoft.DotNet.Arcade.Sdk

* Update dependencies from https://github.com/dotnet/arcade build 20231018.2

Microsoft.DotNet.Arcade.Sdk , Microsoft.DotNet.Helix.Sdk
 From Version 9.0.0-beta.23504.4 -> To Version 9.0.0-beta.23518.2

Dependency coherency updates

Microsoft.DotNet.XliffTasks
 From Version 1.0.0-beta.23502.1 -> To Version 1.0.0-beta.23516.1 (parent: Microsoft.DotNet.Arcade.Sdk

* Update dependencies from https://github.com/dotnet/arcade build 20231018.2

Microsoft.DotNet.Arcade.Sdk , Microsoft.DotNet.Helix.Sdk
 From Version 9.0.0-beta.23504.4 -> To Version 9.0.0-beta.23518.2

Dependency coherency updates

Microsoft.DotNet.XliffTasks
 From Version 1.0.0-beta.23502.1 -> To Version 1.0.0-beta.23516.1 (parent: Microsoft.DotNet.Arcade.Sdk

* Update dependencies from https://github.com/dotnet/arcade build 20231018.2

Microsoft.DotNet.Arcade.Sdk , Microsoft.DotNet.Helix.Sdk
 From Version 9.0.0-beta.23504.4 -> To Version 9.0.0-beta.23518.2

Dependency coherency updates

Microsoft.DotNet.XliffTasks
 From Version 1.0.0-beta.23502.1 -> To Version 1.0.0-beta.23516.1 (parent: Microsoft.DotNet.Arcade.Sdk

* Update dependencies from https://github.com/dotnet/arcade build 20231018.2

Microsoft.DotNet.Arcade.Sdk , Microsoft.DotNet.Helix.Sdk
 From Version 9.0.0-beta.23504.4 -> To Version 9.0.0-beta.23518.2

Dependency coherency updates

Microsoft.DotNet.XliffTasks
 From Version 1.0.0-beta.23502.1 -> To Version 1.0.0-beta.23516.1 (parent: Microsoft.DotNet.Arcade.Sdk

* Update dependencies from https://github.com/dotnet/arcade build 20231018.2

Microsoft.DotNet.Arcade.Sdk , Microsoft.DotNet.Helix.Sdk
 From Version 9.0.0-beta.23504.4 -> To Version 9.0.0-beta.23518.2

Dependency coherency updates

Microsoft.DotNet.XliffTasks
 From Version 1.0.0-beta.23502.1 -> To Version 1.0.0-beta.23516.1 (parent: Microsoft.DotNet.Arcade.Sdk

* Restore .NET SDK using `build.ps1`

* Update toolset version

* Update VS version to 17.8

* Update dependencies from https://github.com/dotnet/arcade build 20231024.1

Microsoft.DotNet.Arcade.Sdk , Microsoft.DotNet.Helix.Sdk
 From Version 9.0.0-beta.23504.4 -> To Version 9.0.0-beta.23524.1

Dependency coherency updates

Microsoft.DotNet.XliffTasks
 From Version 1.0.0-beta.23502.1 -> To Version 1.0.0-beta.23523.1 (parent: Microsoft.DotNet.Arcade.Sdk

* Update dependencies from https://github.com/dotnet/arcade build 20231025.2

Microsoft.DotNet.Arcade.Sdk , Microsoft.DotNet.Helix.Sdk
 From Version 9.0.0-beta.23504.4 -> To Version 9.0.0-beta.23525.2

Dependency coherency updates

Microsoft.DotNet.XliffTasks
 From Version 1.0.0-beta.23502.1 -> To Version 1.0.0-beta.23523.1 (parent: Microsoft.DotNet.Arcade.Sdk

---------

Co-authored-by: dotnet-maestro[bot] <dotnet-maestro[bot]@users.noreply.github.com>
Co-authored-by: Jan Jones <janjones@microsoft.com>

CyrusNajmabadi committed:
Merge remote-tracking branch 'upstream/main' into authorizationException

AlekseyTs committed:
Fix regression in definite assignment analysis for out parameters in presence of local functions. (#70563)

Fixes #69775.

maryamariyan committed:
Apply feedback

ToddGrun committed:
GetEditDistance has various optimizations that are used when a threshold is passed in that weren't being allowed to execute when called from BKTree.Lookup as it needed the precise edit distance. However, this method only needs the precise edit distance if it has edges to later inspect. (#70567)

I tested this out locally adding stopwatch calls in Lookup with both the old and new code and saw a 25-30% reduction in overall time spent in Lookup with this change.

CyrusNajmabadi commented on src/EditorFeatures/Core/Classification/Semantic/ClassificationUtilities.cs:
```diff
@@ -196,12 +196,12 @@ public static SegmentedList<ITagSpan<IClassificationTag>> Convert(IClassificatio
             {
                 await classificationService.AddSemanticClassificationsAsync(
-                   document, snapshotSpan.Span.ToTextSpan(), options, classifiedSpans, cancellationToken).ConfigureAwait(false);
+                   document, new[] { snapshotSpan.Span.ToTextSpan() }, options, classifiedSpans, cancellationToken).ConfigureAwait(false);
```
note: we'll want these to be immutable arrays for the final form.  but for the temp form, arrays are fine.

CyrusNajmabadi commented on src/EditorFeatures/Test2/Classification/ClassificationTests.vb:
```diff
@@ -94,8 +95,9 @@ Namespace Microsoft.CodeAnalysis.Editor.UnitTests.Classification
+                Dim spanArray As TextSpan() = {New TextSpan(0, text.Length)}
                 Dim spans = Await ClassifierHelper.GetClassifiedSpansAsync(
-                    document, New TextSpan(0, text.Length), ClassificationOptions.Default, includeAdditiveSpans:=False, CancellationToken.None)
+                    document, spanArray, ClassificationOptions.Default, includeAdditiveSpans:=False, CancellationToken.None)
```
```suggestion
                    document, {New TextSpan(0, text.Length)}, ClassificationOptions.Default, includeAdditiveSpans:=False, CancellationToken.None)
```

CyrusNajmabadi commented on src/EditorFeatures/Test2/Classification/ClassificationTests.vb:
```diff
@@ -94,8 +95,9 @@ Namespace Microsoft.CodeAnalysis.Editor.UnitTests.Classification
                 Dim document = workspace.CurrentSolution.Projects.Single().Documents.Single()
                 Dim text = Await document.GetTextAsync()
 
+                Dim spanArray As TextSpan() = {New TextSpan(0, text.Length)}
```
```suggestion
```

CyrusNajmabadi commented on src/EditorFeatures/Test2/Classification/ClassificationTests.vb:
```diff
@@ -155,8 +157,9 @@ Namespace Microsoft.CodeAnalysis.Editor.UnitTests.Classification
+                Dim spanArray As TextSpan() = {New TextSpan(0, text.Length)}
                 Dim spans = Await ClassifierHelper.GetClassifiedSpansAsync(
-                    document, New TextSpan(0, text.Length), ClassificationOptions.Default, includeAdditiveSpans:=False, CancellationToken.None)
+                    document, spanArray, ClassificationOptions.Default, includeAdditiveSpans:=False, CancellationToken.None)
```
same feedback.

CyrusNajmabadi replied:
and all the cases below.

also, consider having an extension that does this.  that way you wouldn't have to update these use sites.

CyrusNajmabadi commented on src/EditorFeatures/TestUtilities/Classification/AbstractClassifierTests.cs:
```diff
@@ -274,8 +274,8 @@ protected static async Task<ImmutableArray<ClassifiedSpan>> GetSemanticClassific
-            await service.AddSemanticClassificationsAsync(document, span, options, result, CancellationToken.None);
-            await service.AddEmbeddedLanguageClassificationsAsync(document, span, options, result, CancellationToken.None);
+            await service.AddSemanticClassificationsAsync(document, new[] { span }, options, result, CancellationToken.None);
+            await service.AddEmbeddedLanguageClassificationsAsync(document, new[] { span }, options, result, CancellationToken.None);
```
an extension would make it so you could undo all these cases.

CyrusNajmabadi commented on src/Features/Core/Portable/EmbeddedLanguages/Classification/AbstractEmbeddedLanguageClassificationService.cs:
```diff
@@ -36,22 +36,22 @@ internal abstract class AbstractEmbeddedLanguageClassificationService :
 
         public async Task AddEmbeddedLanguageClassificationsAsync(
-            Document document, TextSpan textSpan, ClassificationOptions options, SegmentedList<ClassifiedSpan> result, CancellationToken cancellationToken)
+            Document document, TextSpan[] textSpans, ClassificationOptions options, SegmentedList<ClassifiedSpan> result, CancellationToken cancellationToken)
```
change to immutablearray in teh final form.

ToddGrun commented on src/Workspaces/Core/Portable/Classification/SyntaxClassification/AbstractSyntaxClassificationService.Worker.cs:
```diff
@@ -153,23 +160,29 @@ private void AddClassification(ClassifiedSpan classification)
+                var classifiers = _getTokenClassifiers(syntax);
+                foreach (var textSpan in textSpans)
+                {
+                    ClassifyStructuredTrivia(leadingTrivia);
```
>ClassifyStructuredTrivia(leadingTrivia); [](http://example.com/codeflow?start=20&length=40)

Did you mean to put the ClassifyStructuredTrivia calls outside the outer loop?

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -22,6 +24,11 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens
 {
     internal class SemanticTokensHelpers
     {
+        private static Random s_random = new Random();
```
fine for experimentation.  not for merging in.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -81,13 +88,29 @@ internal class SemanticTokensHelpers
                 await GetClassifiedSpansForDocumentAsync(
-                    classifiedSpans, document, textSpan, options, cancellationToken).ConfigureAwait(false);
+                    classifiedSpans, document, textSpans, options, cancellationToken).ConfigureAwait(false);
+                s_old.Add(stopwatch2.ElapsedMilliseconds);
```
note: this is not threadsafe, and may corrupt.  should do `lock (s_old) ...` (same with s_new).

CyrusNajmabadi commented on src/Tools/ExternalAccess/FSharp/Internal/Classification/FSharpClassificationService.cs:
```diff
@@ -38,11 +39,15 @@ public void AddLexicalClassifications(SourceText text, TextSpan textSpan, Segmen
+                using var _ = s_listPool.GetPooledObject(out var list);
+                await _service.AddSemanticClassificationsAsync(document, textSpans[0], list, cancellationToken).ConfigureAwait(false);
+                result.AddRange(list);
+            }
```
i would not do it this way.  just loop over each textspan and call into their service, aggregating their results into the result  array.

that way if roslyn ever decides to break up the classification call into small spans, F# will continue to work.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Classification/AbstractClassificationService.cs:
```diff
@@ -159,13 +159,16 @@ private static bool IsFullyLoaded(Document document, CancellationToken cancellat
-                        result.Add(new ClassifiedSpan(span, ClassificationTypeNames.ReassignedVariable));
+                    foreach (var textSpan in textSpans)
+                    {
+                        var reassignedVariableSpans = await reassignedVariableService.GetLocationsAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
```
note: this may end up being a perf hit for razor as well.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Classification/SyntaxClassification/AbstractSyntaxClassificationService.Worker.cs:
```diff
@@ -98,10 +99,12 @@ private void ProcessNodes()
                     _cancellationToken.ThrowIfCancellationRequested();
                     var nodeOrToken = _pendingNodes.Pop();
+                    var intervalsThatIntersect = _textSpanIntervalTree.GetIntervalsThatIntersectWith(
+                            nodeOrToken.FullSpan.Start, nodeOrToken.FullSpan.Length);
```
```suggestion
                        nodeOrToken.FullSpan.Start, nodeOrToken.FullSpan.Length);
```

CyrusNajmabadi replied:
this is costly to do in a loop.  if you actually need to literally pass this data down (unclear if that's needed), then create a temp buffer and call *Add*IntervalsThatIntersectWith.  Then clear the buffer on each loop.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Classification/SyntaxClassification/AbstractSyntaxClassificationService.Worker.cs:
```diff
@@ -56,14 +57,14 @@ internal partial class AbstractSyntaxClassificationService
                 CancellationToken cancellationToken)
             {
-                using var worker = new Worker(semanticModel, textSpan, list, getNodeClassifiers, getTokenClassifiers, options, cancellationToken);
+                using var worker = new Worker(semanticModel, textSpans, list, getNodeClassifiers, getTokenClassifiers, options, cancellationToken);
```
i actually think you do *not* want to do it this way.  INstead, loop over textSpans *here*, and just call the normal 'single text span' code that already exists from that point.

maryamariyan committed:
undo TextSpanIntervalTree for abstract syntax classification

minor:
stop the stopwatch before getting elapsed time

dotnet-bot committed:
Merge pull request #70545 from dotnet/merges/release/dev17.8-to-main

Merge release/dev17.8 to main

CyrusNajmabadi committed:
Merge pull request #70558 from CyrusNajmabadi/authorizationException

Handle authorization exceptions when reading files

maryamariyan committed:
Apply PR feedback - use ImmutableArray

maryamariyan committed:
make AddSyntacticClassifications take multiple spans

maryamariyan committed:
temp for testing
